File: 238. Product of Array Except Self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> prods = vector<int>(nums.size(), 1);
        for (int i = 1; i < nums.size(); i++) {
            prods[i] = prods[i - 1] * nums[i - 1];
        }
        
        for (int j = nums.size() - 1, m = 1; j >=0; j--) {
            prods[j] = prods[j] * m;
            m *= nums[j];
        }
        
        return prods;
    }
};

File: 238. Product of Array Except Self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        # prod[i] = left[i] * right[i]
        left = [1] * len(nums)
        right = [1] * len(nums)
        i = 1
        j = len(nums) - 2
        while i < len(nums):
            left[i] = left[i - 1] * nums[i - 1]
            i += 1
        while j >= 0:
            right[j] = right[j + 1] * nums[j + 1]
            j -= 1
        
        # print(left)
        # print(right)
        ret = []
        for i in range(len(nums)):
            ret.append(left[i] * right[i])
        return ret
